File: 22.백트래킹/15651.N과 M (3)/main.py
import sys

input = sys.stdin.readline


def bt(n: int):
    # 종료 조건
    if n == N:
        print(*arr)
        return

    # 중복 허용 선택
    for i in range(1, M + 1):
        arr[n] = i
        bt(n + 1)


if __name__ == "__main__":
    M, N = map(int, input().split())
    arr = [-1] * N
    bt(0)

File: 22.백트래킹/15651.N과 M (3)/main.ts
const rl = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

const inputs: number[] = [];

rl.on("line", (line: string) => {
  const values = line.split(" ").map(Number);
  inputs.push(...values);
}).on("close", () => {
  const N = inputs[0];
  const M = inputs[1];
  solution(N, M);
  process.exit();
});

function solution(N: number, M: number) {
  function backtrack(n: number) {
    if (n === M) {
      answer.push(arr.join(" "));
      return;
    }

    for (let i = 1; i <= N; i++) {
      arr[n] = i;
      backtrack(n + 1);
    }
  }
  const answer: string[] = [];
  const arr: number[] = [];
  backtrack(0);
  console.log(answer.join("\n"));
}
